src/librustc_mir/transform/add_yk_swt_calls.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::ty::{self, TyCtxt, List};
+use rustc::mir::{Operand, LocalDecl, Place, SourceInfo, BasicBlock, Local, BasicBlockData,
+    TerminatorKind, Terminator, OUTERMOST_SOURCE_SCOPE, Constant, Mir};
+use rustc_data_structures::indexed_vec::Idx;
+use syntax_pos::DUMMY_SP;
+use syntax::attr;
+use transform::{MirPass, MirSource};
+use rustc::hir;
+use rustc::hir::def_id::{DefIndex, LOCAL_CRATE};
+use rustc::hir::map::blocks::FnLikeNode;
+
+/// A MIR pass which, for each basic block, inserts a call to the software trace recorder.
+/// The call arguments passed uniquely identify the MIR location.
+pub struct AddYkSWTCalls(pub DefIndex);
+
+impl MirPass for AddYkSWTCalls {
+    fn run_pass<'a, 'tcx>(&self,
+                          tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                          src: MirSource,
+                          mir: &mut Mir<'tcx>) {
+        if !should_annotate(tcx, src) {
+            return;
+        }
+
+        // Find the recorder function to call.
+        let rec_fn_defid = tcx.get_lang_items(LOCAL_CRATE).yk_swt_rec_loc_wrap()
+            .expect("couldn't find software trace recorder function");
+
+        // Types.
+        let unit_ty = tcx.mk_unit();
+        let u32_ty = tcx.types.u32;
+        let u64_ty = tcx.types.u64;
+
+        // Each block is replaced by a new block whose terminator calls the recorder function.
+        let mut replace_blocks = Vec::new();
+
+        // The original blocks are copied and the recorder function returns to a copy.
```

Comment:
Not sure I understand this comment.

Reply:
We are replacing each block with one that makes a call to the tracing function before returning to a copy of the original block. We have to do it this way to preserve the edge structure of the CFG.

Reply:
Right, so it might be worth pulling these comments out to be something more like:

```
// In order to perform software tracing, we replace each MIR block with a new block
// that first records the block as having been traced and then calls the original block.
```

then I think the two variables are fairly obvious.

Reply:
I was going to suggest ASCII art, but thought better of it! Let's give it a shot.